File: python_mockup/elevator.py
DEFAULT_HEIGHT = 34
DEFAULT_FLOOR = 0
DEFAULT_DIRECTION = True #heading up from the ground floor
DEFAULT_USERS = {} 

class Elevator(object): 
    
    def __init__(self, direction=DEFAULT_DIRECTION, height=DEFAULT_HEIGHT,
                 floor=DEFAULT_FLOOR, users=DEFAULT_USERS):
        self.direction = direction
        self.height = height
        self.floor = floor
        self.users = users

    def __repr__(self):
        msg = "Direction: {direction:<2} Height: {height:<3} "\
              "Floor:{floor:<2} Users: {users}"
        return msg.format(direction=self.direction, height=self.height,
                          floor=self.floor, users=self.users)

    def moveFloor(self):
        usersToRemove = []
        increment = 1 if self.direction else -1
        self.floor += increment

        for userId, user in self.users.items():
            user.moveFloor()
            user.currentFloor = self.floor
            if user.hasReachedDestination():
                usersToRemove.append(userId)

        for userId in usersToRemove:
            self.users[userId].addBoardingDuration()
            del self.users[userId]

    def userEnters(self, user):
        user.addBoardingDuration()
        self.users[user.id] = user

    def hasUsersRiding(self):
        return len(self.users) == 0

File: Elevator.cpp
#include "stdafx.h"
#include "Elevator.h"


const int DEFAULT_HEIGHT = 34;
const int DEFAULT_FLOOR = 0;
const int DEFAULT_DIRECTION = true; // heading up

Elevator::Elevator(){
	height = DEFAULT_HEIGHT;
	floor = DEFAULT_FLOOR;
	direction = DEFAULT_DIRECTION;
}
Elevator::Elevator(int h, int f, bool d) {
	height = h;
	floor = f;
	direction = d;
}
int Elevator::getHeight(){return height;}
int Elevator::getFloor(){return floor;}
bool Elevator::getDirrection(){return direction;}
bool Elevator::hasUsersRiding(){return users.size() == 0;}
void Elevator::moveFloor(){
	vector<string> usersToRemove;
	int increment;
	if (direction)
		increment = 1;
	else
		increment = -1;
	floor += increment;
	for (map<string, User>::iterator itr = users.begin(); itr != users.end(); itr++)
	{
		itr->second.moveFloor();
		itr->second.setCurrentFloor(floor);
		if (itr->second.hasReachedDestination())
		{
			usersToRemove.push_back(itr->second.getUserId());
		}
	}
	for (int i = 0; i < usersToRemove.size(); i++){
		users[usersToRemove[i]].addBoardingDuration();
		users.erase(usersToRemove[i]);
	}
}
void Elevator::userEnters(User usr){
	usr.addBoardingDuration();
	users[usr.getUserId()] = usr;
}
void Elevator::setHeight(int newHeight){height = newHeight;}
void Elevator::setFloor(int newFloor){floor = newFloor;}
void Elevator::setDirection(bool newDir){direction = newDir;}
Elevator::~Elevator(){}
